### DISCUSSION TITLE
Extra detail for callHierarchy

### DISCUSSION THREAD

**Discussion by nagi1999a** [clangd/clangd#1940] [Ideas]
Recently I started to use clangd to help developing, and find it extremely great!
However, in VScode, I found that when showing call hierarchy, there is no extra information about each function that showed at the right of function name, which utilized the `detail` field in the responded LSP json.
![image](https://github.com/clangd/clangd/assets/49529145/56c62355-2c2b-4f58-9ab1-f18c1cb5f8ed)
Most extensions in vscode utilized the field to show the file name and/or the line number information, and we can also make it possible to show file name by simply add one line to the [clang-tools-extra/clangd/XRefs.cpp](https://github.com/llvm/llvm-project/blob/02e17ab1b97a8c0dc22facc8c66850e5aca28b60/clang-tools-extra/clangd/XRefs.cpp)

```
diff --git a/clang-tools-extra/clangd/XRefs.cpp b/clang-tools-extra/clangd/XRefs.cpp
index cd9092664..a6f0dcbcd 100644
--- a/clang-tools-extra/clangd/XRefs.cpp
+++ b/clang-tools-extra/clangd/XRefs.cpp
@@ -1757,6 +1757,7 @@ static std::optional<HierarchyItem> symbolToHierarchyItem(const Symbol &S,
   // (https://github.com/clangd/clangd/issues/59).
   HI.range = HI.selectionRange;
   HI.uri = Loc->uri;
+  HI.detail = llvm::sys::path::filename(Loc->uri.file()).str();
 
   return HI;
 }
```

And then we can see the changes in the VScode:
![image](https://github.com/clangd/clangd/assets/49529145/0d5501f5-d588-40d1-9328-83bf2818456e)

It might be great that we can customize the information by some arguments of clangd. Also, current implementation seems to not split each call within a single function, maybe we can also have arguments to control the behavior, too.

Thanks again for the great tool!

**Comment by HighCommander4** [comment]
> we can also make it possible to show function name by simply add one line to the [clang-tools-extra/clangd/XRefs.cpp](https://github.com/llvm/llvm-project/blob/02e17ab1b97a8c0dc22facc8c66850e5aca28b60/clang-tools-extra/clangd/XRefs.cpp)

I think you meant to say "to show the **file** name", right?

Showing the file name sounds useful to me. I do have one concern about it: in #915, it was proposed that the `"detail"` field should be used for the function signature, which also seems to be useful. However, we only have one `"detail"` field.

Do you think we should somehow combine multiple pieces of information (e.g. file name **and** function signature) into the detail field? Or just pick one of them?

**Reply by nagi1999a** [reply]
> I think you meant to say "to show the file name", right?

Yes, thanks for quick reply and sorry for the mistake!

> Do you think we should somehow combine multiple pieces of information (e.g. file name and function signature) into the detail field? Or just pick one of them?

I think we can choose one of them as the default behavior, and switch using yaml configuration, like `filename | signature | all`, because it seems that either case are useful in different places.
In addition, for the `all` option, we may provide json type results, to make other extensions easily parse them.

**Reply by HighCommander4** [reply]
> I think we can choose one of them as the default behavior, and switch using yaml configuration, like `filename | signature | all`, because it seems that either case are useful in different places.

A couple of additional things I realized:

 * For methods, it's also important to show the class name.
 * There is a [patch](https://github.com/llvm/llvm-project/pull/77556) under review adding support for outgoing calls, which actually populates the `detail` field with the function's fully qualified name.
    * This includes the class name for methods, and any namespaces the function or class is contained in. It does also repeat the function name itself, which is a bit redundant; maybe that could be removed.

And I've looked at what some other language servers do:

 * typescript: shows class name (does not repeat function name), does not show signature
 * rust-analyzer: shows signature (the full declaration including the function name), does not show class name

I'm not opposed to having a config option governing this personally, though I'd want to get some feedback from other maintainers on that (some have expressed a preference for being judicious with the use of config options). Either way it's important to have a good default.

> In addition, for the `all` option, we may provide json type results, to make other extensions easily parse them.

Could you clarify what you mean by extensions here -- do you mean editor extensions which are language clients, like vscode-clangd?

**Reply by nagi1999a** [reply]
Thanks for the detail survey on this topic!

It seems that different language server have distinct implementations! It is also interesting that, in vscode, official extensions for python and javascript do not show extra information in call hierarchy. And the official extension for c/cpp, [vscode-cpptools](https://github.com/microsoft/vscode-cpptools), and some 3rd party ones, [c-hierarchy](https://github.com/0w0miki/c-hierarchy) & [VSCode-C-Call-Hierarchy](https://github.com/abdalmoniem/VSCode-C-Call-Hierarchy), tends to show the filename and line number. 

> Could you clarify what you mean by extensions here -- do you mean editor extensions which are language clients, like vscode-clangd?

Yes, extensions may caught LSP json and do some post-processing before actually send to the editor, they might provide advanced mixed format for different purposes, like "<filename>@<line number>" or "<signature>\n<filename>", but will require extra effort for each extension to modify to intercept and forward LSP messages.

**Reply by HighCommander4** [reply]
Another idea to consider could be to use a textual separator like "•" to combine multiple things into `detail`.

For example: classname • signature • filename

The "classname" part would only need to be included for C++ methods.

The "signature" we might also limit to C++, since in C functions cannot be overloaded. (And maybe even in C++ we could show the signature only for functions which are actually overloaded, though that might require extra index lookups to determine.)

That would leave just the filename in C, which is what you originally asked for.

**Reply by travkin79** [reply]
I'd like to add a few considerations to this discussion.

In the Eclipse IDE, LSP-based editors use the project / plug-in [LSP4E](https://github.com/eclipse/lsp4e). LSP4E provides an LSP client implementation that is intended to be used for all kinds of language servers. Here, they seem to comply with VSCode's implementation and do the following:
   - They assume the `name` field to contain either a function's name or a container name (i.e. a class name) and a function name separated by a colon (:), e.g. `MyClass:myFunction()`. I think, the function name might include a function signature.
   - The `detail` field is assumed to be empty or to contain e.g. the file name.
   - The call hierarchy GUI uses labels built of all three parts (see [implementation](https://github.com/eclipse/lsp4e/blob/68f9534c97107529e94184cc864917da1f910551/org.eclipse.lsp4e/src/org/eclipse/lsp4e/callhierarchy/CallHierarchyLabelProvider.java#L41)), i.e. something like "Container:function() file"

For me, it's even more important to have the functions' containing type (class) in some of the [CallHierarchyItem](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#callHierarchyItem) fields, in best case complying with the LSP4E and VSCode implementation.

**Comment by i-garrison** [comment]
Thanks @HighCommander4 for pointing me here!

Is it possible to extend `details` field into something like property map with fixed language-specific keys like `signature` and `return value` etc.? This way we could add everything of interest to c++.

I tried to modify both XFefs.cpp to provide function signature in details field and LSP4E to use that to show in call hierarchy. While this change kind of worked (for c++, except for constructors in my tests) I'm not sure if using single details field would be handy to pass everything of interest. Consider adding function return type, would require some kind of encoded representation so IDE could decode and properly represent both function signature and return value type.

**Reply by travkin79** [reply]
Hi @i-garrison,
I think, it's not possible to extend the `detail` field into a property map (without having to encode some kind of key-value pairs) since the LSP specification defined the `detail` field as a plain string (see [specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#callHierarchyItem)). Besides, changing the `detail` field's content for _clangd_ and adapting LSP4E to read it might break LSP4E for other languages / language servers. But there is another field called `data` that might be used for that purpose since it could be of any type. I don't know how it is used today by _clangd_.

**Reply by HighCommander4** [reply]
The `data` field is used by clangd to store a stable identifier for the call hierarchy item between related LSP requests (e.g. a `prepareCallHierarchy` request and a subsequent `incomingCalls` request). This avoids clangd having to re-identify the item in its internal model based on the other properties of the `CallHierarchyItem` (e.g. by doing a name lookup on `name` and using `uri` and `range` to disambiguate between symbols with the same name). The value currently stored is a string key into clangd's internal symbol table.

From the client's point of view, the value of `data` is an opaque value that it's expected to pass back to the server.

I think it would be possible to reuse `data` to also store other properties of interest to the client as an extension. Since the LSP specifies its type as `LSPAny`, we could change it from a string to an object with multiple fields. One of them, let's say `id`, would be the string key that's stored today for re-identification purposes, but other fields like `signature`, `returnValue`, or `scope` could be added if desired.

Clients which aren't aware of this extension would just send back this entire object without inspecting it. (It's a bit wasteful to be sending back the fields other than `id`, but it's not a big deal.) Clients which _are_ aware of this extension can examine the object and use the `signature` etc. fields to enhance their rendering of the hierarchy.

**Reply by i-garrison** [reply]
Can we consider extending `CallHierarchyItem` with extra `LanguageSpecific` `LSPArray` of language-specific property objects? I understand extending protocol is more work but would prefer not abusing `data` field for feature probably useful beyond `C` and `C++` and maybe even `Java`.

**Reply by HighCommander4** [reply]
@i-garrison Yes, I think we could do that, maybe initially as an extension; see https://reviews.llvm.org/D137894 for a previous case where we added a new field to an LSP type (in that case, `Location`) as an extension.

We could also propose the new field as an official part of the LSP itself; that would involve filing an issue [here](https://github.com/microsoft/language-server-protocol/issues/).

**Reply by i-garrison** [reply]
> @i-garrison Yes, I think we could do that, maybe initially as an extension; see https://reviews.llvm.org/D137894 for a previous case where we added a new field to an LSP type (in that case, `Location`) as an extension.
> 
> We could also propose the new field as an official part of the LSP itself; that would involve filing an issue [here](https://github.com/microsoft/language-server-protocol/issues/).

Hi @HighCommander4 I just filled a proposal https://github.com/microsoft/language-server-protocol/issues/2118 with what I can imagine right away. Please fill free to copy/edit/re-propose as you see fit.

**Comment by travkin79** [comment]
Hi @HighCommander4,
according to your [comment](https://github.com/clangd/clangd/issues/1185#issuecomment-2689642955) from issue #1185 the call hierarchy will contain the class name (beginning with clangd vers. 20). That are great news! Thank you very much!

Will there also be some information about the parent file or the method signature? Which fields will contain these details?

**Reply by HighCommander4** [reply]
> Will there also be some information about the parent file or the method signature?

The change in clangd 20 only added the class name for methods. It was mostly an incidental result of adding support for outgoing calls (the original patch by Quentin Chateau to implement outgoing calls, which I adopted and got across the finish line, contained this change regarding class names in code that was common to incoming and outgoing calls).

I do think the other additions (signature and file name) would be valuable, they just haven't been implemented yet. (Note that for function signature, the implementation requires storing this data in the index as discussed in #915.)

**Reply by travkin79** [reply]
Hi @HighCommander4,
I'm experimenting with clangd 20 using it in Eclipse IDE through [LSP4E ](https://github.com/eclipse/lsp4e) and [CDT LSP](https://github.com/eclipse-cdt/cdt-lsp) and I'm facing a few minor issues with the `name` and `detail` fields in `CallHierarchyItem` coming from the language server / clangd. It seems, LSP4E expects slightly other values here.

Could we adapt clangd to comply with what LSP4E (and VSCode, a reference implementation) expect? Please note, LSP4E is implementing support for arbitrary language servers and expects them to behave in the same way. (I could create a clangd GitHub issue for that if needed.)

The results from LSP4E look something like the following.

![image](https://github.com/user-attachments/assets/b975269f-7b41-49ef-9c26-8e8b3470d075)

- The label for the containing element (class) also contains the method's name.
- The function in my example is missing its containing namespace.

LSP4E seems to use the same algorithm as (the de-facto standard) VSCode for creating the call hierarchy items' labels and relies on a certain format in the two fields.

- The `name` field is expected to have the format `<method name>|<method name>:<return type>`. clangd does not return the optional return type. But I can live with that.
- The `detail` field is expected to contain the container's fully qualified name, something like `[<namespace>::]<class name>`. clangd does also append the method name. I'd like to remove the method name from this field.

**Reply by i-garrison** [reply]
Hi @travkin79 I think I failed to find description of the de-facto standard used by VSCode, could you please provide a link?

**Reply by travkin79** [reply]
Hi @i-garrison,
I did only find a hint in the LSP4E source code that [mentioned VSCode as de-facto standard](https://github.com/eclipse-lsp4e/lsp4e/blob/6e4750f1534b3b0d1700bc5367510119eeac6c5b/org.eclipse.lsp4e/src/org/eclipse/lsp4e/callhierarchy/CallHierarchyLabelProvider.java#L65). But it seems, the LSP specification maintainers [do really prefer using VSCode as a reference implementation](https://github.com/microsoft/language-server-protocol/pull/2003#issuecomment-2287942018).

**Reply by i-garrison** [reply]
> Hi @i-garrison, I did only find a hint in the LSP4E source code that [mentioned VSCode as de-facto standard](https://github.com/eclipse-lsp4e/lsp4e/blob/6e4750f1534b3b0d1700bc5367510119eeac6c5b/org.eclipse.lsp4e/src/org/eclipse/lsp4e/callhierarchy/CallHierarchyLabelProvider.java#L65). But it seems, the LSP specification maintainers [do really prefer using VSCode as a reference implementation](https://github.com/microsoft/language-server-protocol/pull/2003#issuecomment-2287942018).

Thanks, looks like this extension idea is probably not needed to help with proper CDT call hierarchy then https://github.com/clangd/clangd/discussions/1940#discussioncomment-11397939 though it could be a bit more flexible.

**Reply by HighCommander4** [reply]
@travkin79 Removing the method name appearing in duplicate in the `detail` field sounds totally reasonable to me. Please go ahead and file a clangd issue about that.

For the return type, my preference would be for clangd to pioneer a protocol extension along the lines @i-garrison suggested, to store additional fields like the return type, signature / parameter types, and filename in a structured fashion. LSP4E could early-adopt such an extension without waiting for it to be standardized and obtain the return type from there for servers that support it.

**Reply by travkin79** [reply]
Thank you @HighCommander4,
I created issue #2346 to remove the method name from the `detail` field.

Using additional fields for the method signature sounds reasonable to me. I think, I can support you with early-adopting LSP4E to use them.

At the moment, I'm stuck with another [clangd PR](https://github.com/llvm/llvm-project/pull/113669) that I'd like to early-adopt in LSP4E. Someone has to finish the work. It seems there is not much work left, but the original developer seems not to be available anymore.
